internal/provider/providers/myaddr/provider.go

```diff
@@ -0,0 +1,178 @@
+
+	// handle HTTP response
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// key is valid, decode response to get name
+		var respBody struct {
+			Name string `json:"name"`
+		}
+		if err = json.NewDecoder(resp.Body).Decode(&respBody); err != nil {
+			err = fmt.Errorf("initializing provider (myaddr): json decoding http response: %w", err)
+			return
+		}
+		p.name = respBody.Name
+	case http.StatusBadRequest, http.StatusNotFound:
+		// key is invalid
+		err = fmt.Errorf("initializing provider (myaddr): %w", errors.ErrKeyNotValid)
+	default:
+		// unexpected HTTP status
+		err = fmt.Errorf("initializing provider (myaddr): %w: %d: %s",
+			errors.ErrHTTPStatusNotValid, resp.StatusCode, utils.BodyToSingleLine(resp.Body))
+	}
+	return
+}
+
+func (p *Provider) String() string {
+	return utils.ToString(p.Domain(), p.Owner(), constants.Myaddr, p.IPVersion())
+}
+
+func (p *Provider) Domain() string {
+	return p.name + ".myaddr.tools"
+}
+
+func (p *Provider) Owner() string {
+	return "@"
+}
+
+func (p *Provider) BuildDomainName() string {
+	return p.Domain()
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.Domain(), p.Domain()),
+		Owner:     p.Owner(),
+		Provider:  "<a href=\"https://myaddr.tools/\">myaddr</a>",
+		IPVersion: p.IPVersion().String(),
+	}
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) IPv6Suffix() netip.Prefix {
+	return p.ipv6Suffix
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+	// generate HTTP request
+	u := &url.URL{
+		Scheme: "https",
+		Host:   "myaddr.tools",
+		Path:   "/update",
+	}
+	v := url.Values{}
+	v.Set("key", p.key)
+	v.Set("ip", ip.String())
+	b := strings.NewReader(v.Encode())
+	var req *http.Request
+	if req, err = http.NewRequestWithContext(ctx, http.MethodPost, u.String(), b); err != nil {
+		err = fmt.Errorf("creating http request: %w", err)
+		return
+	}
+	headers.SetContentType(req, "application/x-www-form-urlencoded")
+	headers.SetUserAgent(req)
+
+	// execute HTTP request
+	var resp *http.Response
+	if resp, err = client.Do(req); err != nil {
+		return
+	}
+	defer resp.Body.Close()
+	defer io.Copy(io.Discard, resp.Body) // ensure body is read to completion
+
+	// handle HTTP response
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// update was successful
+		newIP = ip
+	case http.StatusBadRequest:
+		// bad request
+		err = fmt.Errorf("%w: %s", errors.ErrBadRequest, utils.BodyToSingleLine(resp.Body))
+	case http.StatusNotFound:
+		// registration not found
+		err = fmt.Errorf("%w: %s", errors.ErrKeyNotValid, utils.BodyToSingleLine(resp.Body))
+	default:
+		// unexpected HTTP status
+		err = fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, resp.StatusCode, utils.BodyToSingleLine(resp.Body))
+	}
+	return
```

Comment:
- some unneeded comments
- no naked returns
- return early, almost always, easier to read

```suggestion
	switch response.StatusCode {
	case http.StatusOK:
		return ip, nil
	case http.StatusBadRequest:
		return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrBadRequest, utils.BodyToSingleLine(response.Body))
	case http.StatusNotFound
		return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrKeyNotValid, utils.BodyToSingleLine(response.Body))
	default:
		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
	}
```